File: VECTOAux/VectoAuxiliaries/Hvac/SSMTOOL.vb
Imports Omu.ValueInjecter
Imports VectoAuxiliaries.Hvac
Imports Newtonsoft.Json
Imports System.IO
Imports System.Reflection
Imports System.Text
Imports TUGraz.VectoCommon.Utils

Namespace Hvac
	'Used by frmHVACTool
	'Replaces Spreadsheet model which does the same calculation
	'Version of which appears on the form title.
	Public Class SSMTOOL
		Implements ISSMTOOL

		Private filePath As String
		Public Property GenInputs As ISSMGenInputs Implements ISSMTOOL.GenInputs
		Public Property TechList As ISSMTechList Implements ISSMTOOL.TechList
		Public Property Calculate As ISSMCalculate Implements ISSMTOOL.Calculate
		Public Property SSMDisabled As Boolean Implements ISSMTOOL.SSMDisabled
		Public Property HVACConstants As IHVACConstants Implements ISSMTOOL.HVACConstants

		'Repeat Warning Flags
		Private CompressorCapacityInsufficientWarned As Boolean
		Private FuelFiredHeaterInsufficientWarned As Boolean

		'Base Values
		Public ReadOnly Property ElectricalWBase As Double Implements ISSMTOOL.ElectricalWBase
			Get
				Return If(SSMDisabled, 0, Calculate.ElectricalWBase()) '.SI(Of Watt)()
			End Get
		End Property

		Public ReadOnly Property MechanicalWBase As Double Implements ISSMTOOL.MechanicalWBase
			Get
				Return If(SSMDisabled, 0, Calculate.MechanicalWBase) '.SI(Of Watt)()
			End Get
		End Property

		Public ReadOnly Property FuelPerHBase As Double Implements ISSMTOOL.FuelPerHBase
			Get
				Return If(SSMDisabled, 0, Calculate.FuelPerHBase())	'.SI(Of LiterPerHour)()
			End Get
		End Property

		'Adjusted Values
		Public ReadOnly Property ElectricalWAdjusted As Double Implements ISSMTOOL.ElectricalWAdjusted
			Get
				Return If(SSMDisabled, 0, Calculate.ElectricalWAdjusted()) '.SI(Of Watt)()
			End Get
		End Property

		Public ReadOnly Property MechanicalWBaseAdjusted As Double Implements ISSMTOOL.MechanicalWBaseAdjusted
			Get
				Dim mechAdjusted As Double = If(SSMDisabled, 0, Calculate.MechanicalWBaseAdjusted())

				If _
					CompressorCapacityInsufficientWarned = False AndAlso
					(mechAdjusted) / (1000 * GenInputs.AC_COP) > GenInputs.AC_CompressorCapacitykW Then

					OnMessage(Me,
							"HVAC SSM :AC-Compressor Capacity unable to service cooling, run continues as if capacity was sufficient.",
							AdvancedAuxiliaryMessageType.Warning)
					CompressorCapacityInsufficientWarned = True

				End If


				Return mechAdjusted	' .SI(Of Watt)()
			End Get
		End Property

		Public ReadOnly Property FuelPerHBaseAdjusted As Double Implements ISSMTOOL.FuelPerHBaseAdjusted
			Get
				Return If(SSMDisabled, 0, Calculate.FuelPerHBaseAdjusted())	' .SI(Of LiterPerHour)()
			End Get
		End Property

		'Constructors
		Sub New(filePath As String, hvacConstants As HVACConstants, Optional isDisabled As Boolean = False,
				Optional useTestValues As Boolean = False)

			Me.filePath = filePath
			Me.SSMDisabled = isDisabled
			Me.HVACConstants = hvacConstants

			GenInputs = New SSMGenInputs(useTestValues, fPATH(filePath))
			TechList = New SSMTechList(filePath, GenInputs, useTestValues)

			Calculate = New SSMCalculate(Me)
		End Sub

		'Clone values from another object of same type
		Public Sub Clone(from As ISSMTOOL) Implements ISSMTOOL.Clone

			Dim feedback As String = String.Empty

			GenInputs.InjectFrom(DirectCast(from, SSMTOOL).GenInputs)

			TechList.Clear()

			For Each line As TechListBenefitLine In DirectCast(from, SSMTOOL).TechList.TechLines

				Dim newLine As New TechListBenefitLine(Me.GenInputs)
				newLine.InjectFrom()
				newLine.InjectFrom(line)
				TechList.Add(newLine, feedback)

			Next
		End Sub

		'Persistance Functions
		Public Function Save(filePath As String) As Boolean Implements ISSMTOOL.Save


			Dim returnValue As Boolean = True
			Dim settings As JsonSerializerSettings = New JsonSerializerSettings()
			settings.TypeNameHandling = TypeNameHandling.Objects

			'JSON METHOD
			Try

				Dim output As String = JsonConvert.SerializeObject(Me, Formatting.Indented, settings)

				File.WriteAllText(filePath, output)

			Catch ex As Exception

				'Nothing to do except return false.
				returnValue = False

			End Try

			Return returnValue
		End Function

		Public Function Load(filePath As String) As Boolean Implements ISSMTOOL.Load

			Dim returnValue As Boolean = True
			Dim settings As JsonSerializerSettings = New JsonSerializerSettings()
			Dim tmpAux As SSMTOOL ' = New SSMTOOL(filePath, HVACConstants)

			settings.TypeNameHandling = TypeNameHandling.Objects

			'JSON METHOD
			Try

				Dim output As String = File.ReadAllText(filePath)


				tmpAux = JsonConvert.DeserializeObject(Of SSMTOOL)(output, settings)

				tmpAux.TechList.SetSSMGeneralInputs(tmpAux.GenInputs)

				For Each tll As TechListBenefitLine In tmpAux.TechList.TechLines

					tll.inputSheet = tmpAux.GenInputs

				Next


				'This is where we Assume values of loaded( Deserialized ) object.
				Clone(tmpAux)

			Catch ex As Exception

				'Nothing to do except return false.

				returnValue = False
			End Try

			Return returnValue
		End Function

		'Comparison
		Public Function IsEqualTo(source As ISSMTOOL) As Boolean Implements ISSMTOOL.IsEqualTo

			'In this methods we only want to compare the non Static , non readonly public properties of 
			'The class's General, User Inputs and  Tech Benefit members.

			Return compareGenUserInputs(source) AndAlso compareTechListBenefitLines(source)
		End Function

		Private Function compareGenUserInputs(source As ISSMTOOL) As Boolean

			Dim src As SSMTOOL = DirectCast(source, SSMTOOL)

			Dim returnValue As Boolean = True

			Dim properties As PropertyInfo() = Me.GenInputs.GetType.GetProperties

			For Each prop As PropertyInfo In properties

				'If Not prop.GetAccessors.IsReadOnly Then
				If prop.CanWrite Then
					If Not prop.GetValue(Me.GenInputs, Nothing).Equals(prop.GetValue(src.GenInputs, Nothing)) Then
						returnValue = False
					End If

				End If

			Next

			Return returnValue
		End Function

		Private Function compareTechListBenefitLines(source As ISSMTOOL) As Boolean


			Dim src As SSMTOOL = DirectCast(source, SSMTOOL)

			'Equal numbers of lines check
			If Me.TechList.TechLines.Count <> src.TechList.TechLines.Count Then Return False

			For Each tl As ITechListBenefitLine In _
				Me.TechList.TechLines.OrderBy(Function(o) o.Category).ThenBy(Function(n) n.BenefitName)

				'First Check line exists in other
				If _
					src.TechList.TechLines.Where(Function(w) w.BenefitName = tl.BenefitName AndAlso w.Category = tl.Category).Count <>
					1 Then

					Return False
				Else

					'check are equal

					Dim testLine As ITechListBenefitLine =
							src.TechList.TechLines.First(Function(w) w.BenefitName = tl.BenefitName AndAlso w.Category = tl.Category)

					If Not testLine.IsEqualTo(tl) Then
						Return False
					End If

				End If


			Next

			'All Looks OK
			Return True
		End Function

		'Overrides
		Public Overrides Function ToString() As String


			Dim sb As New StringBuilder

			sb.AppendLine(Calculate.ToString())


			Return sb.ToString()
		End Function

		'Dynamicly Get Fuel having re-adjusted Engine Heat Waste, this was originally supposed to be Solid State. Late adjustment request 24/3/2015
		Public Function FuelPerHBaseAsjusted(AverageUseableEngineWasteHeatKW As Double) As Double _
			Implements ISSMTOOL.FuelPerHBaseAsjusted

			If SSMDisabled Then
				Return 0
			End If

			'Set Engine Waste Heat
			GenInputs.AH_EngineWasteHeatkW = AverageUseableEngineWasteHeatKW
			Dim fba As Double = FuelPerHBaseAdjusted

			'Dim FuelFiredWarning As Boolean = fba * GenInputs.BC_AuxHeaterEfficiency * HVACConstants.FuelDensity * GenInputs.BC_GCVDieselOrHeatingOil * 1000 > (AverageUseableEngineWasteHeatKW + GenInputs.AH_FuelFiredHeaterkW)

			'If Not FuelFiredHeaterInsufficientWarned AndAlso FuelFiredWarning Then

			'    FuelFiredHeaterInsufficientWarned = True

			'    OnMessage(Me, " HVAC SSM : Fuel fired heater insufficient for heating requirements, run continues assuming it was sufficient.", AdvancedAuxiliaryMessageType.Warning)

			'End If

			Return fba
		End Function

		'Events
		Public Event Message(ByRef sender As Object, message As String, messageType As AdvancedAuxiliaryMessageType) _
			Implements ISSMTOOL.Message

		'Raise Message Event.
		Private Sub OnMessage(sender As Object, message As String, messageType As AdvancedAuxiliaryMessageType)


			If Not message Is Nothing Then

				RaiseEvent Message(Me, message, messageType)

			End If
		End Sub
	End Class
End Namespace

File: VECTOAux/BusAuxiliaries/DownstreamModules/Impl/HVAC/SSMTOOL.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using Newtonsoft.Json;
using Omu.ValueInjecter;
using TUGraz.VectoCore.BusAuxiliaries.Interfaces;
using TUGraz.VectoCore.BusAuxiliaries.Interfaces.DownstreamModules.HVAC;
using TUGraz.VectoCore.BusAuxiliaries.Util;

namespace TUGraz.VectoCore.BusAuxiliaries.DownstreamModules.Impl.HVAC
{
	// Used by frmHVACTool
	// Replaces Spreadsheet model which does the same calculation
	// Version of which appears on the form title.
	public class SSMTOOL : ISSMTOOL
	{
		private string filePath;
		public ISSMGenInputs GenInputs { get; set; }
		public ISSMTechList TechList { get; set; }
		public ISSMCalculate Calculate { get; set; }
		public bool SSMDisabled { get; set; }
		public IHVACConstants HVACConstants { get; set; }

		// Repeat Warning Flags
		private bool CompressorCapacityInsufficientWarned;
		private bool FuelFiredHeaterInsufficientWarned;

		// Base Values
		public double ElectricalWBase
		{
			get {
				return SSMDisabled ? 0 : Calculate.ElectricalWBase; // .SI(Of Watt)()
			}
		}

		public double MechanicalWBase
		{
			get {
				return SSMDisabled ? 0 : Calculate.MechanicalWBase; // .SI(Of Watt)()
			}
		}

		public double FuelPerHBase
		{
			get {
				return SSMDisabled ? 0 : Calculate.FuelPerHBase; // .SI(Of LiterPerHour)()
			}
		}

		// Adjusted Values
		public double ElectricalWAdjusted
		{
			get {
				return SSMDisabled ? 0 : Calculate.ElectricalWAdjusted; // .SI(Of Watt)()
			}
		}

		public double MechanicalWBaseAdjusted
		{
			get {
				var mechAdjusted = SSMDisabled ? 0 : Calculate.MechanicalWBaseAdjusted;

				if (CompressorCapacityInsufficientWarned == false && (mechAdjusted) / (1000 * GenInputs.AC_COP) > GenInputs.AC_CompressorCapacitykW) {
					OnMessage(this, "HVAC SSM :AC-Compressor Capacity unable to service cooling, run continues as if capacity was sufficient.", AdvancedAuxiliaryMessageType.Warning);
					CompressorCapacityInsufficientWarned = true;
				}


				return mechAdjusted; // .SI(Of Watt)()
			}
		}

		public double FuelPerHBaseAdjusted
		{
			get {
				return SSMDisabled ? 0 : Calculate.FuelPerHBaseAdjusted; // .SI(Of LiterPerHour)()
			}
		}

		// Constructors
		public SSMTOOL(string filePath, HVACConstants hvacConstants, bool isDisabled = false, bool useTestValues = false)
		{
			this.filePath = filePath;
			this.SSMDisabled = isDisabled;
			this.HVACConstants = hvacConstants;

			GenInputs = new SSMGenInputs(useTestValues, FilePathUtils.fPATH(filePath));
			TechList = new SSMTechList(filePath, GenInputs, useTestValues);

			Calculate = new SSMCalculate(this);
		}

		// Clone values from another object of same type
		public void Clone(ISSMTOOL from)
		{
			var feedback = string.Empty;

			GenInputs.InjectFrom((SSMTOOL)from.GenInputs);

			TechList.Clear();

			foreach (var line in from.TechList.TechLines) {
				var newLine = new TechListBenefitLine(this.GenInputs);
				// newLine.InjectFrom()
				newLine.InjectFrom(line);
				TechList.Add(newLine, ref feedback);
			}
		}

		// Persistance Functions
		public bool Save(string filePath)
		{
			var returnValue = true;
			var settings = new JsonSerializerSettings();
			settings.TypeNameHandling = TypeNameHandling.Objects;

			// JSON METHOD
			try {
				var output = JsonConvert.SerializeObject(this, Formatting.Indented, settings);

				File.WriteAllText(filePath, output);
			} catch (Exception ex) {

				// Nothing to do except return false.
				returnValue = false;
			}

			return returnValue;
		}

		public bool Load(string filePath)
		{
			var returnValue = true;
			var settings = new JsonSerializerSettings();
			SSMTOOL tmpAux; // = New SSMTOOL(filePath, HVACConstants)

			settings.TypeNameHandling = TypeNameHandling.Objects;

			// JSON METHOD
			try {
				var output = File.ReadAllText(filePath);


				tmpAux = JsonConvert.DeserializeObject<SSMTOOL>(output, settings);

				tmpAux.TechList.SetSSMGeneralInputs(tmpAux.GenInputs);

				foreach (TechListBenefitLine tll in tmpAux.TechList.TechLines)

					tll.inputSheet = tmpAux.GenInputs;


				// This is where we Assume values of loaded( Deserialized ) object.
				Clone(tmpAux);
			} catch (Exception ex) {

				// Nothing to do except return false.

				returnValue = false;
			}

			return returnValue;
		}

		// Comparison
		public bool IsEqualTo(ISSMTOOL source)
		{

			// In this methods we only want to compare the non Static , non readonly public properties of 
			// The class's General, User Inputs and  Tech Benefit members.

			return compareGenUserInputs(source) && compareTechListBenefitLines(source);
		}

		private bool compareGenUserInputs(ISSMTOOL source)
		{
			var src = (SSMTOOL)source;

			var returnValue = true;

			var properties = this.GenInputs.GetType().GetProperties();

			foreach (var prop in properties) {

				// If Not prop.GetAccessors.IsReadOnly Then
				if (prop.CanWrite) {
					if (!prop.GetValue(this.GenInputs, null/* TODO Change to default(_) if this is not a reference type */).Equals(prop.GetValue(src.GenInputs, null/* TODO Change to default(_) if this is not a reference type */)))
						returnValue = false;
				}
			}

			return returnValue;
		}

		private bool compareTechListBenefitLines(ISSMTOOL source)
		{
			var src = (SSMTOOL)source;

			// Equal numbers of lines check
			if (this.TechList.TechLines.Count != src.TechList.TechLines.Count)
				return false;

			foreach (var tl in this.TechList.TechLines.OrderBy(o => o.Category).ThenBy(n => n.BenefitName)) {

				// First Check line exists in other
				if (src.TechList.TechLines.Where(w => w.BenefitName == tl.BenefitName && w.Category == tl.Category).Count() != 1)
					return false;
				else {

					// check are equal

					var testLine = src.TechList.TechLines.First(w => w.BenefitName == tl.BenefitName && w.Category == tl.Category);

					if (!testLine.IsEqualTo(tl))
						return false;
				}
			}

			// All Looks OK
			return true;
		}

		// Overrides
		public override string ToString()
		{
			var sb = new StringBuilder();

			sb.AppendLine(Calculate.ToString());


			return sb.ToString();
		}

		// Dynamicly Get Fuel having re-adjusted Engine Heat Waste, this was originally supposed to be Solid State. Late adjustment request 24/3/2015
		public double FuelPerHBaseAsjusted(double AverageUseableEngineWasteHeatKW)
		{
			if (SSMDisabled)
				return 0;

			// Set Engine Waste Heat
			GenInputs.AH_EngineWasteHeatkW = AverageUseableEngineWasteHeatKW;
			var fba = FuelPerHBaseAdjusted;

			// Dim FuelFiredWarning As Boolean = fba * GenInputs.BC_AuxHeaterEfficiency * HVACConstants.FuelDensity * GenInputs.BC_GCVDieselOrHeatingOil * 1000 > (AverageUseableEngineWasteHeatKW + GenInputs.AH_FuelFiredHeaterkW)

			// If Not FuelFiredHeaterInsufficientWarned AndAlso FuelFiredWarning Then

			// FuelFiredHeaterInsufficientWarned = True

			// OnMessage(Me, " HVAC SSM : Fuel fired heater insufficient for heating requirements, run continues assuming it was sufficient.", AdvancedAuxiliaryMessageType.Warning)

			// End If

			return fba;
		}

		// Events
		public event MessageEventHandler Message;

		// Raise Message Event.
		private void OnMessage(object sender, string message, AdvancedAuxiliaryMessageType messageType)
		{
			if (message != null) {
				object ssmtool = this;
				Message?.Invoke(ref ssmtool, message: message, messageType: messageType);
			}
		}
	}
}
